### ISSUE TITLE
TLS Cipher 3b: record: use PSA keys

### ISSUE THREAD

**Issue by mpg** [ARMmbed/mbedtls#5206]
Use PSA-based keys and operations for TLS record en/decryption  when `MBEDTLS_USE_PSA_CRYPTO` is enabled.

- [x] In `mbedtls_ssl_encrypt_buf()` replace calls to `mbedtls_cipher_crypt()` and `mbedtls_cipher_auth_encrypt_ext()` with calls to `psa_cipher_encrypt()` and `psa_aead_encrypt()`.
- [x] In `mbedtls_ssl_encrypt_buf()` replace calls to `mbedtls_cipher_crypt()` and `mbedtls_cipher_auth_encrypt_ext()` with calls to `psa_cipher_encrypt()` and `psa_aead_encrypt()`.

Depends on: #5204 
Follow-up: #5207

**Comment by mprse** [comment]
I understand that the second bullet should refer to decryption functions (not encryption).

Seem that `mbedtls_cipher_crypt()` can't be replaced with `psa_cipher_encrypt()/psa_cipher_decrypt()` as these functions do not provide IV and the following sequence needs to be used: `psa_cipher_(en/de)crypt_setup(), psa_cipher_set_iv(), psa_cipher_update(), psa_cipher_finish()`.

**Comment by mprse** [comment]
@mpg @gilles-peskine-arm I created a PR https://github.com/ARMmbed/mbedtls/pull/5434 that implements PSA encryption/decryption for TLS Cipher (still code needs to be cleaned up). 
Now I can see now that a lot of tests are failing on CI. I started analyzing `test_siute_ssl` build using `scripts/config.py full` config and I need some help.

I can see that `Record crypt` tests are are failing. This test is not ready for PSA crypto. `ssl_tls12_populate_transform()/ssl_tls13_populate_transform()` functions are not used in this test (to import keys in PSA style) instead test has its own function `build_transforms()`. I modified this function to init PSA and import psa keys and set alg when PSA is enabled. After this some tests have passed.
Next test that failed was `Record crypt, AES-192-CBC, 1.2, SHA-384` and it failed because translation function (https://github.com/ARMmbed/mbedtls/issues/5182) does not support `MBEDTLS_CIPHER_AES_192_CBC`.

I'm confused and not sure if I'm going in the right direction.
How this test should be adapted to PSA? If cipher type is not supported then test should be skipped?

**Comment by gilles-peskine-arm** [comment]
@mprse I don't quite understand your question. I see that for the “Record crypt” tests, in `mbedtls_ssl_encrypt_buf`, `psa_cipher_encrypt_setup` is called with `transform->psa_alg == 0`. So it looks like `psa_alg` should have been set but wasn't.

**Comment by mpg** [comment]
It seems weird to me that we have TLS tests for AES-192-CBC as there is no ciphersuite that uses that.

I think there's a bit of a mismatch between (1) the strategy that was used for these tests, which apparently was to test with every cipher we support in the cipher layer, and (2) the strategy used in `mbedtls_cipher_to_psa()` (as defined in #5434) which was to focus only on what's used in TLS.

We could resolve that either by expanding `mbedtls_cipher_to_psa()` or by removing the tests that don't relate to an existing ciphersuite. I'm inclined to remove the tests, because the alternative means adding code size to the library for no apparent purpose.

@hanno-arm Since you originally added the "Record crypt" tests, including those for AES-192-CBC, can you clarify if there was a specific reason for including ciphers that are not used by any TLS ciphersuite, or if it was more "doesn't cost anything, and more testing can't hurt"?

**Comment by mpg** [comment]
> `mbedtls_cipher_to_psa` should support all algorithms that are supported by both sides. It isn't specific to TLS.

Well, in the current version of #5434 which introduces it, it's declared in `ssl_misc.h` and implemented in `ssl_tls.c`.

I'm not sure in what other place we would use it: X.509 doesn't use symmetric ciphers, and for `mbedtls_cipher_setup_psa()` we use different translation functions (which I didn't want to extend, because it would increase the range of what `cipher_setup_psa()` supports, meaning we'd have to extend its testing, which IMO is not where we want to be spending our time, considering we're moving away from `cipher_setup_psa()` and I even hope to [deprecate it](https://github.com/ARMmbed/mbedtls/issues/5261) soonish).

Again, there's two ways to resolve this: either rename to clarify that it's TLS-specific and perhaps add a word about that in the documentation, or extend to support everything and move to `psa_util.h` (which doesn't have a corresponding .c file, so only contains `static inline` functions, which I find suboptimal for functions of this size, but that's a separate issue).

**Comment by mprse** [comment]
> @mprse I don't quite understand your question. I see that as of 474ca2a, for the “Record crypt” tests, in mbedtls_ssl_encrypt_buf, psa_cipher_encrypt_setup is called with transform->psa_alg == 0. So it looks like psa_alg should have been set but wasn't.

PSA alg and keys are set in `ssl_tls12_populate_transform()/ssl_tls13_populate_transform()` functions according to the design and these functions are not used by Record crypt test function. So yes, `psa_alg` should have been set but wasn't. My question is how this test should be adapted?
On my side (locally) I modified `build_transforms()` test function to init PSA alg and keys when PSA crypto is enabled.

> Again, there's two ways to resolve this: either rename to clarify that it's TLS-specific and perhaps add a word about that in the documentation, or extend to support everything and move to psa_util.h (which doesn't have a corresponding .c file, so only contains static inline functions, which I find suboptimal for functions of this size, but that's a separate issue).

Regarding `mbedtls_cipher_to_psa()` maybe if PSA is enabled and translation function can't do the conversion like in case of `AES-192-CBC` the test should be skipped? Then we can rework `mbedtls_cipher_to_psa()` (extend and move to other location).

**Comment by mpg** [comment]
> On my side (locally) I modified `build_transforms()` test function to init PSA alg and keys when PSA crypto is enabled.

I think that's the right way to do it.

> Regarding `mbedtls_cipher_to_psa()` maybe if PSA is enabled and translation function can't do the conversion like in case of `AES-192-CBC` the test should be skipped?

That's indeed an option, but I'd rather remove the tests if they aren't useful (which I don't think they are).

**Comment by mprse** [comment]
> That's indeed an option, but I'd rather remove the tests if they aren't useful (which I don't think they are).

If the tests aren't useful I think we can remove them. Can I assume that `ciphers that are not used by any TLS ciphersuite` are the ones that can't be converted using `mbedtls_cipher_to_psa()`.

> rename to clarify that it's TLS-specific
`mbedtls_cipher_to_psa_tls` ?

**Comment by mpg** [comment]
> Can I assume that `ciphers that are not used by any TLS ciphersuite` are the ones that can't be converted using `mbedtls_cipher_to_psa()`.

You can work based on that assumption, and I'll double-check that you didn't remove too much when reviewing the PR.

> `mbedtls_cipher_to_psa_tls` ?

Let's go with `mbedtls_ssl_cipher_to_psa()` instead, following the pattern of the existing TLS-specific functions.
